kbp/config/acl_checker.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package config
+
+import (
+	"errors"
+	"path"
+	"strings"
+)
+
+type permissionsV1 struct {
+	read bool
+	list bool
+}
+
+// username -> permissionsV1
+type accessControlV1 struct {
+	whitelistAdditional map[string]permissionsV1
+	anonymous           permissionsV1
+}
+
+func parsePermissionsV1(permsStr string) (permissionsV1, error) {
+	permsStr = strings.TrimSpace(permsStr)
+	var perms permissionsV1
+	if len(permsStr) == 0 {
+		return perms, nil
+	}
+	for _, p := range strings.Split(permsStr, ",") {
+		switch p {
+		case "read":
+			perms.read = true
+		case "list":
+			perms.list = true
+		default:
+			return permissionsV1{}, ErrInvalidPermissions{
+				permissions: permsStr}
+		}
+	}
+	return perms, nil
+}
+
+func (ac *AccessControlV1) makeAccessControlV1Internal(users map[string]UserV1) (
+	aci *accessControlV1, err error) {
+	if ac == nil {
+		return nil, errors.New("nil AccessControlV1")
+	}
+	aci = &accessControlV1{}
+	aci.anonymous, err = parsePermissionsV1(ac.AnonymousPermissions)
+	if err != nil {
+		return nil, err
+	}
+	for username, permissions := range ac.WhitelistAdditionalPermissions {
+		if _, ok := users[username]; !ok {
+			return nil, ErrUndefinedUsername{username: username}
+		}
+		if aci.whitelistAdditional == nil {
+			aci.whitelistAdditional = make(map[string]permissionsV1)
+		}
+		aci.whitelistAdditional[username], err = parsePermissionsV1(permissions)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return aci, nil
+}
+
+type aclCheckerV1 struct {
+	children map[string]*aclCheckerV1
+	aci      *accessControlV1
+}
+
+// cleanPath cleans thePath in following order:
+//   1) Call path.Clean;
+//   2) Remove leading and trailing "/";
+//   3) Split the rest.
+// If thePath is root, nil elems is returned. Otherwise, elems is guaranteed to
+// be non-empty.
+//
+// If split2 is true, at most elems has at most 2 elements and the second one
+// would include everything other than the first element in the path;
+// otherwise, the entire thePath is split by "/" till the end of the world.
+func cleanPath(thePath string, split2 bool) (elems []string) {
```

Comment:
"thePath" -> "p" would be better in my opinion.  ("path" would be better but it conflicts.  I just really don't like putting "the" in var names.)